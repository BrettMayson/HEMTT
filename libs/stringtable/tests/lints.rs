#![allow(clippy::unwrap_used)]

use hemtt_stringtable::{
    analyze::{lint_all, lint_one},
    Project,
};
use hemtt_workspace::{
    reporting::{Codes, WorkspaceFiles},
    LayerType,
};

const ROOT: &str = "tests/lints/";

macro_rules! lint {
    ($dir:ident) => {
        paste::paste! {
            #[test]
            fn [<simple_ $dir>]() {
                insta::assert_snapshot!(lint(stringify!($dir)));
            }
        }
    };
}

lint!(l01_sorted);

fn lint(file: &str) -> String {
    let folder = std::path::PathBuf::from(ROOT);
    let workspace = hemtt_workspace::Workspace::builder()
        .physical(&folder, LayerType::Source)
        .finish(None, false, &hemtt_common::config::PDriveOption::Disallow)
        .unwrap();
    let source = workspace.join(format!("{file}.xml")).unwrap();
    let workspace_files = WorkspaceFiles::new();
    let stringtable = Project::read(source).unwrap();

    let mut codes: Codes = Vec::new();
    codes.extend(lint_one(&stringtable, None, vec![]));
    codes.extend(lint_all(&vec![stringtable], None, vec![]));

    codes.retain(|e| e.ident().starts_with(&format!("L-{}", file.split_once('_').unwrap().0.to_uppercase())));

    codes
        .iter()
        .map(|e| e.diagnostic().unwrap().to_string(&workspace_files))
        .collect::<Vec<_>>()
        .join("\n")
        .replace('\r', "")
}
