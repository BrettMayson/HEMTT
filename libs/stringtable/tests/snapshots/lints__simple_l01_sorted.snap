---
source: libs/stringtable/tests/lints.rs
expression: lint(stringify! (l01_sorted))
---
[0m[1m[38;5;11mwarning[L-L01][0m[1m: Stringtable at `` is not sorted[0m
 [0m[36m=[0m [33mhelp[0m: Run `hemtt ln sort` to sort the stringtable
